rfc/system/3844-artifact-service.md

```diff
@@ -0,0 +1,148 @@
+
+Basic functionality here on the search side might look like
+* Given an Artifact URL, list all executions it was used in.
+* Given an Artifact URL, find the execution that produced it, or let the user know if it was an upload.
+* Given a set of aliases or tags, return Artifact objects.
+
+On the authoring side, we would like users to be able to do the following
+```python
+@task
+def t1() -> Annotated[nn.Module, Artifact(name="models.nn.lidar", alias="latest", overwrite_alias=True)]: ...
+
+# Possibly in a completely different repo
+@workflow
+def wf(model: nn.Module = Artifact.get_query(name="models.nn.lidar", alias="latest")): ...
+```
+The binding is made at compilation time but it is not resolved until run time. That is, when `wf` is run, Admin will have to call out to the Artifact service, get back precisely one Artifact, and use the Literal therein contained as the input. If the type is not a match, then it can result it in a user run time error, which is okay.
+
+## 3 Proposed Implementation
+We are working through an implementation in the following PRs:
+IDL: https://github.com/flyteorg/flyteidl/pull/408/files
+Admin: https://github.com/flyteorg/flyteadmin/pull/572/files
+Flytekit: https://github.com/flyteorg/flytekit/pull/1655/files
+
+This Flytekit [PR](https://github.com/flyteorg/flytekit/pull/1674/files) was also referenced earlier and will be done as a separate initiative to fix long-standing issues with `pyflyte run`
+
+There is also considerable work left to be done on the UI side. If the tiny URL construct holds it would be nice to see a click-copying button from the UI for example that would copy a Python snippet that would allow users to immediately hydrate a Literal.
+
+## 4 Metrics & Dashboards
+Will address last.
+
+
+## 5 Drawbacks
+None.
+
+## 6 Alternatives
+One alternative not to the idea presented here but to the implementation is to use the existing data catalog service instead, either in whole or in part. Everything from using the grpc service itself with well-crafted calls, to implementing the new service but with the data catalog code.
+
+## 7 Potential Impact and Dependencies
+Everything will be implemented before merging anything major or anything that can't be pulled back out.
+
+## 8 Unresolved questions
+If you look in the IDL change, you will see that the Artifact message comes with a `source` field. This is nice but doesn't really make sense if you think about it. There's a better name for the act of identifying where an Artifact came from, and also for identifying things in the search capability like what other executions it was used in. That term is lineage. We're not sure if lineage should be a different service or is an integral part of this Artifact service yet. I can see arguments both ways. It is clear that they are closely intertwined though.
```

Comment:
My personal opinion is that this should be a lineage service and not a separate one.